src/main/java/org/opentripplanner/model/plan/legreference/LegReferenceSerializer.java

```diff
@@ -0,0 +1,99 @@
+package org.opentripplanner.model.plan.legreference;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.text.ParseException;
+import java.util.Base64;
+import javax.annotation.Nullable;
+import org.opentripplanner.model.FeedScopedId;
+import org.opentripplanner.model.calendar.ServiceDate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Serializer for LegReferences
+ */
+public class LegReferenceSerializer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LegReferenceSerializer.class);
+
+    /** private constructor to prevent instantiating this utility class */
+    private LegReferenceSerializer() {}
+
+
+    @Nullable
+    public static String encode(LegReference legReference) {
+        if (legReference == null) { return null; }
+        LegReferenceType typeEnum = LegReferenceType.forClass(legReference.getClass());
+
+        if (typeEnum == null) {
+            throw new IllegalArgumentException("Unknown LegReference type");
+        }
+
+        var buf = new ByteArrayOutputStream();
+        try(var out = new ObjectOutputStream(buf)){
+            // The order must be the same in the encode and decode function
+            writeEnum(typeEnum, out);
+            typeEnum.getSerializer().write(legReference, out);
+            out.flush();
+            return Base64.getUrlEncoder().encodeToString(buf.toByteArray());
+        }
+        catch (IOException e) {
+            LOG.error("Failed to encode leg reference", e);
+            return null;
+        }
+    }
+
+    @Nullable
+    public static LegReference decode(String legReference) {
+        if (legReference == null) { return null; }
+
+        var buf = Base64.getUrlDecoder().decode(legReference);
+        var input = new ByteArrayInputStream(buf);
+
+        try(var in = new ObjectInputStream(input)) {
+            // The order must be the same in the encode and decode function
+
+            var type = readEnum(in, LegReferenceType.class);
+            return type.getDeserializer().read(in);
+        }
+        catch (IOException | ParseException e) {
+            LOG.error("Unable to decode leg reference: '" + legReference + "'", e);
+            return null;
+        }
+    }
+
+    static void writeScheduledTransitLeg(LegReference ref, ObjectOutputStream out) throws IOException {
+        if (ref instanceof ScheduledTransitLegReference s) {
+            out.writeUTF(s.tripId().toString());
```

Comment:
I couldn't immediately understand from the Java documentation how this works. How are the various parts separated from each other? Does ObjectOutputStream do it automatically?

Reply:
I find the LegReference concept very interesting since some of my clients would like to have a feature to completely restore an itinerary including realtime updates.

The use case is the following: you buy a ticket for a particular itinerary which is saved on your phone. On the day of the travel you open the app and want to see the exact itinerary that you booked but also want to see the realtime updates for it.

Why am I talking about this? I think for this feature we want to future-proof those references a tiny bit. What to you think about including a version number in the encoded form?

Reply:
Yes, `ObjectOutputStream` is responsible for writing the values to the buffer, the various parts are not separated from each other, but instead it is the responsibility of the implementer to write and read them in the same order.

This feature can support that functionality, at the moment for only scheduled transit leg, but is is possible to create references to other types of legs as well. When we have an implementation for all types, it would be possible to concatenate the legs in order to create a reference to a complete itinerary. Already now, you could store references to all transit legs, and use them to fetch the status of each trip.

The `LegReferenceType` already has a version in the type name. That way we can have an individual version number for each type of leg.